cmd/natc/ippool/consensusippool.go

```diff
@@ -0,0 +1,435 @@
+		return whereWhen{}, false
+	}
+	ww, ok := ps.addrToDomain.Load(addr)
+	if !ok {
+		log.Printf("domainLookup: peer state doesn't recognize addr: %s", addr)
+		return whereWhen{}, false
+	}
+	return ww, true
+}
+
+// StartConsensus is part of the IPPool interface. It starts the raft background routines that handle consensus.
+func (ipp *ConsensusIPPool) StartConsensus(ctx context.Context, ts *tsnet.Server, clusterTag string) error {
+	cfg := tsconsensus.DefaultConfig()
+	cfg.ServeDebugMonitor = true
+	cns, err := tsconsensus.Start(ctx, ts, ipp, clusterTag, cfg)
+	if err != nil {
+		return err
+	}
+	ipp.consensus = cns
+	return nil
+}
+
+type whereWhen struct {
+	Domain   string
+	LastUsed time.Time
+}
+
+type consensusPerPeerState struct {
+	domainToAddr map[string]netip.Addr
+	addrToDomain *syncs.Map[netip.Addr, whereWhen]
+}
+
+// StopConsensus is part of the IPPool interface. It stops the raft background routines that handle consensus.
+func (ipp *ConsensusIPPool) StopConsensus(ctx context.Context) error {
+	return (ipp.consensus).(*tsconsensus.Consensus).Stop(ctx)
+}
+
+// unusedIPV4 finds the next unused or expired IP address in the pool.
+// IP addresses in the pool should be reused if they haven't been used for some period of time.
+// reuseDeadline is the time before which addresses are considered to be expired.
+// So if addresses are being reused after they haven't been used for 24 hours say, reuseDeadline
+// would be 24 hours ago.
+func (ps *consensusPerPeerState) unusedIPV4(ipset *netipx.IPSet, reuseDeadline time.Time) (netip.Addr, bool, string, error) {
+	// If we want to have a random IP choice behavior we could make that work with the state machine by doing something like
+	// passing the randomly chosen IP into the state machine call (so replaying logs would still be deterministic).
+	for _, r := range ipset.Ranges() {
+		ip := r.From()
+		toIP := r.To()
+		if !ip.IsValid() || !toIP.IsValid() {
+			continue
+		}
+		for toIP.Compare(ip) != -1 {
+			ww, ok := ps.addrToDomain.Load(ip)
+			if !ok {
+				return ip, false, "", nil
+			}
+			if ww.LastUsed.Before(reuseDeadline) {
+				return ip, true, ww.Domain, nil
+			}
+			ip = ip.Next()
+		}
+	}
+	return netip.Addr{}, false, "", errors.New("ip pool exhausted")
+}
+
+// isCloseToExpiry returns true if the lastUsed and now times are more than
+// half the lifetime apart
+func isCloseToExpiry(lastUsed, now time.Time, lifetime time.Duration) bool {
+	return now.Sub(lastUsed).Abs() > (lifetime / 2)
+}
+
+type readDomainForIPArgs struct {
+	NodeID tailcfg.NodeID
+	Addr   netip.Addr
+}
+
+// executeReadDomainForIP parses a readDomainForIP log entry and applies it.
+func (ipp *ConsensusIPPool) executeReadDomainForIP(bs []byte) tsconsensus.CommandResult {
+	var args readDomainForIPArgs
+	err := json.Unmarshal(bs, &args)
+	if err != nil {
+		return tsconsensus.CommandResult{Err: err}
+	}
+	return ipp.applyReadDomainForIP(args.NodeID, args.Addr)
+}
+
+// applyReadDomainForIP TODO fran
```

Comment:
i think this is fine you can drop the comment for now, it buddies up with the above easily